### ISSUE TITLE
엔티티의 생명주기

### ISSUE THREAD

**Issue by huajin7282** [SpringStudyInf/spring-is-comming#46]
변경 감지와 병합에 대한 강의를 듣다가 영속 엔티티, 준영속 엔티티 개념이 너무 어려워서 정리해보았다.

엔티티의 생명주기는 아래와 같이 4단계의 생명 주기가 있다.
_비영속(new/transient)
영속(managed)
준영속(detached)
삭제(removed)_

먼저 강의에서 자주 다루어진 **영속**은 **영속성 컨텍스트에 의해서 관리되어지는 상태**이다.
생서한 객체를 em.persist()를 통해 영속성 컨텍스트에 저장하거나, em.find()를 통해 DB에서 엔티티를 조회하면 영속성 컨텍스트에 저장되어 관리되어지므로 영속 상태가 된다.

**준영속**은 영속성 컨텍스트에 저장되었다가 분리된 상태로 **현재는 영속 상태가 아닌 상태**이다.
```
public String updateItem(@ModelAttribute("form") BookForm form) {
 Book book = new Book();
 book.setId(form.getId());
 book.setName(form.getName());
 book.setPrice(form.getPrice());
 book.setStockQuantity(form.getStockQuantity());
 book.setAuthor(form.getAuthor());
 book.setIsbn(form.getIsbn());
 itemService.saveItem(book);
 return "redirect:/items";
 }
```
그런데 강의를 듣다 보니 이해가 가지 않았던 부분은 위의 Book 객체는 이미 DB에 한번 저장되서 준영속 객체라는 설명이다.  new Book() 하고나서 persist()를 한적도 없고 find()를 한적도 없는데 왜 준영속 상태라는 건지 이해하기 어려워서 검색해 보았더니,
준영속의 핵심은 식별자를 기준으로 영속상태가 되어서 DB에 저장된 적이 있는지 라고 한다.
Book 객체의 경우 수정을 위해서 html form에 데이터를 노출한 이후에 다시 new로 재조립된 엔티티이기 때문에 준영속 엔티티가 된다. 심지어 다른 원격지 서버에 엔티티를 네트워크로 전송할 경우에도 원격지 서버에 도착한 엔티티는 영속성 컨텍스트에서 관리할 수 없기 때문에 준영속 상태라고 한다. 그래서 em.detach() 뿐만 아니라 그냥 JPA의 관리를 넘어선 영역이 되면 다 준영속 상태라고 이해했다.

**비영속**은 **객체를 새로 생성한 상태**이다. JPA와 전혀 관계없이 객체만 생성한 상태라고 한다. 그런데 여기서 또 궁금증이 생겼다. 그럼 준영속이랑 비영속은 둘다 JPA에 관리받고 있지 않은 상태의 엔티티인데 왜 준영속과 비영속이라는 말이 따로 있을까 찾아보았다. 둘의 차이점은 영속상태가 되어본 경험이라고 한다. 즉 준영속 엔티티는 식별자가 존재한다는 것이 보장되지만 비영속은 없을수도 있다. 그래서 merge() 동작에 차이가 있는데 비영속 객체는 merge()를 호출할때 완전히 새로운 엔티티를 만들고 준영속 객체는 DB에서 기존 엔티티를 찾고 그 값을 준영속 상태의 객체로 변경한 후에 반환한다고 한다.

마지막으로 삭제는 캐시와 데이터베이스에서 모두 삭제된 상태이다.

참고
https://www.inflearn.com/questions/70393/book-%EA%B0%9D%EC%B2%B4%EA%B0%80-%EC%99%9C-%EC%A4%80%EC%98%81%EC%86%8D%EC%9D%B8%EA%B2%83%EC%9D%B8%EA%B0%80
https://ttl-blog.tistory.com/109
